src/Reference/Reference.php

```diff
@@ -71,20 +63,13 @@ public function getTitle(): string
      * @param string $string
      *
      * @return string
+     *
+     * @deprecated Use TextNormalizer::normalize() instead
      */
     public static function normalizeReference(string $string): string
     {
-        // Collapse internal whitespace to single space and remove
-        // leading/trailing whitespace
-        $string = \preg_replace('/\s+/', ' ', \trim($string));
-
-        if (!\defined('MB_CASE_FOLD')) {
-            // We're not on a version of PHP (7.3+) which has this feature
-            $string = \str_replace(self::$caseFold[0], self::$caseFold[1], $string);
-
-            return \mb_strtoupper($string, 'UTF-8');
-        }
+        @trigger_error(sprintf('%s::normlizeReference() is deprecated; use %s::normalize() instead', self::class, TextNormalizer::class), E_USER_DEPRECATED);
 
-        return \mb_convert_case($string, \MB_CASE_FOLD, 'UTF-8');
+        return (new TextNormalizer())->normalize($string);
```

Comment:
Considering that these don't actually need class instantiation (new), wouldn't making the public methods static make more sense here? In fact, that's what the deprecated message even says: `Use TextNormalizer::normalize() instead`.

Reply:
> Considering that these don't actually need class instantiation (new), wouldn't making the public methods static make more sense here?

We need to keep non-static methods so they can be used in configuration options, but I don't see why we couldn't have both a static and non-static method to cater to both use cases.  Give me a few minutes and I'll push up this change :)

> In fact, that's what the deprecated message even says: Use TextNormalizer::normalize() instead.

I'm just following the common PHP convention of referencing all class methods using `::`, even the non-static ones.

Reply:
Added via 25382406

Reply:
> We need to keep non-static methods so they can be used in configuration options

I'm at a loss for why instantiating classes into instances is even necessary. What does it bring to the table?

Configurations can easily be adjusted to account for static classes:

```php
'slug_normalizer' => '\\League\\CommonMark\\Normalizer\\SlugNormalizer\\SlugNormalizer(),
```

You can still keep interface support and strict adherence, just using `is_subclass_of` instead of `instanceof`.

FWIW, even if you instantiate a class into an instance, it would still be possible for the class instances to invoke its static methods using either the `->` notation or the `::`, you don't have to create instance method wrappers that invoke the static "varient" i.e.:

```php
$result1 = $instance->staticMethod();
$result2 = $instance::staticMethod();
$this-assertSame($result1, $result2);
```
We do it all the time in Drupal.

Reply:
The more I think about this, the more I'm against using static methods here.  I think I might revert 2538240.

Static methods make sense for internal uses where we won't be swapping out implementations, like within `Reference`.  But when it comes to things like `slug_normalizer`, I think it's important that we require non-static implementations.

For example, somebody may want to swap out our naive `SlugNormalizer` with https://github.com/cocur/slugify.  That library is not used statically and incurs some overhead when instantiating and configuring it, so we should allow users to reuse an already-instantiated object.  Sure, those users could leverage the singleton pattern to allow static calls to this non-static instance, but that's not a practice I want to encourage.

But going back to the static usages, I think we can avoid them entirely by moving the responsibility for normalizing text out of the `Reference` object and into whatever instantiates them (currently these are all parser and listener instances).  This is probably a better refactoring than making everything static or allowing both static and non-static methods.

> FWIW, even if you instantiate a class into an instance, it would still be possible for the class instances to invoke its static methods using either the -> notation or the ::, you don't have to create instance method wrappers that invoke the static "varient"

TIL!  That would definitely work here if we wanted to go static, but again, the more I think about it, the less I feel like static methods are the best choice.

Let me play with both ideas (yours and mine) to see how they look in practice.  Both approaches have their merits, and once I've tinkered with both I'll have a better idea of whether my argument here is valid :wink:

Reply:
Upon further review, it seems that normalizing the `$label` in the `Reference` constructor is unnecessary because the `ReferenceMap` is where that normalization is critical, and it already normalizes the label in each of its methods.  Furthermore, there is no code anywhere in this library that depends on `Reference` containing a normalized label.

Therefore, the only calls to normalizers are happening within other instantiated objects that make repeated calls to the normalizer.  This eliminates the compelling use case to make them statically callable.

I've therefore dropped 2538240 from this PR and removed normalization from `Reference::__construct()`.

Reply:
> For example, somebody may want to swap out our naive `SlugNormalizer` with https://github.com/cocur/slugify. That library is not used statically and incurs some overhead when instantiating and configuring it, so we should allow users to reuse an already-instantiated object. Sure, those users could leverage the singleton pattern to allow static calls to this non-static instance, but that's not a practice I want to encourage.

I think I pretty much rubber-ducked myself while replying to this 😃  Basically, I think I've come to the same conclusion as what you stated above. Regardless of where an external library may be implemented, any external library instance that is setup/configured would be inaccessible from a static method. With that in mind, I think its safe to say that this is done.